prov/rxm/src/rxm_conn.c

```diff
@@ -624,26 +624,37 @@ static void *rxm_conn_eq_read(void *arg)
 static void *rxm_conn_progress(void *arg)
 {
 	struct rxm_ep *rxm_ep = container_of(arg, struct rxm_ep, util_ep);
-	struct rxm_msg_eq_entry entry;
+	struct rxm_msg_eq_entry *entry;
 	int ret;
 
+	entry = calloc(1, RXM_MSG_EQ_ENTRY_SZ);
+	if (!entry) {
+		FI_WARN(&rxm_prov, FI_LOG_EP_CTRL,
+			"Unable to allocate memory!\n");
+		return NULL;
+	}
 	FI_DBG(&rxm_prov, FI_LOG_EP_CTRL, "Starting conn event handler\n");
+
 	while (1) {
-		memset(&entry, 0, sizeof(entry));
-		entry.rd = rxm_eq_sread(rxm_ep, RXM_CM_ENTRY_SZ, &entry);
-		if (entry.rd < 0 && entry.rd != -FI_ECONNREFUSED)
-			return NULL;
+		entry->rd = rxm_eq_sread(rxm_ep, RXM_CM_ENTRY_SZ, entry);
+		if (entry->rd < 0 && entry->rd != -FI_ECONNREFUSED)
+			goto exit;
 
-		if (entry.event == FI_NOTIFY &&
+		if (entry->event == FI_NOTIFY &&
 		    (enum ofi_cmap_signal)((struct fi_eq_entry *)
-					   &entry.cm_entry)->data == OFI_CMAP_EXIT) {
-			FI_DBG(&rxm_prov, FI_LOG_EP_CTRL, "Closing CM thread\n");
-			return NULL;
+					   &entry->cm_entry)->data == OFI_CMAP_EXIT) {
+			FI_DBG(&rxm_prov, FI_LOG_EP_CTRL,
+			       "Closing CM thread\n");
+			goto exit;
 		}
-		ret = rxm_conn_handle_event(rxm_ep, &entry);
+		ret = rxm_conn_handle_event(rxm_ep, entry);
 		if (ret)
-			return NULL;
+			goto exit;
+		memset(entry, 0, RXM_MSG_EQ_ENTRY_SZ);
```

Comment:
I know the original code did a memset to 0, but why? Are we not trusting sread() to be correct?

Reply:
@dmitrygladkov should probably answer this. I left it since the memset was being done within the loop originally. Looking now, it does look like rxm_eq_sread() could set error fields in the non fi_eq_cm_entry part of the overall structure, and that those would not necessarily be reset by subsequent calls without an error.